types/context.go

```diff
@@ -16,167 +13,84 @@ import (
-func NewContext(ms MultiStore, header abci.Header, isCheckTx bool, logger log.Logger) Context {
-	c := Context{
-		Context: context.Background(),
-		pst:     newThePast(),
-		gen:     0,
-	}
-
-	c = c.WithMultiStore(ms)
-	c = c.WithBlockHeader(header)
-	c = c.WithChainID(header.ChainID)
-	c = c.WithIsCheckTx(isCheckTx)
-	c = c.WithTxBytes(nil)
-	c = c.WithLogger(logger)
-	c = c.WithVoteInfos(nil)
-	c = c.WithGasMeter(stypes.NewInfiniteGasMeter())
-	c = c.WithMinGasPrices(DecCoins{})
-	c = c.WithConsensusParams(nil)
-	c = c.WithEventManager(NewEventManager())
-
-	return c
-}
-
-// is context nil
-func (c Context) IsZero() bool {
-	return c.Context == nil
-}
-
-// ----------------------------------------------------------------------------
-// Getters
-// ----------------------------------------------------------------------------
-
-type contextKey int // local to the context module
-
-const (
-	contextKeyMultiStore contextKey = iota
-	contextKeyBlockHeader
-	contextKeyChainID
-	contextKeyIsCheckTx
-	contextKeyTxBytes
-	contextKeyLogger
-	contextKeyVoteInfos
-	contextKeyGasMeter
-	contextKeyBlockGasMeter
-	contextKeyMinGasPrices
-	contextKeyConsensusParams
-	contextKeyEventManager
-)
-
-// context value for the provided key
-func (c Context) Value(key interface{}) interface{} {
-	value := c.Context.Value(key)
-	if cloner, ok := value.(cloner); ok {
-		return cloner.Clone()
-	}
-	if message, ok := value.(proto.Message); ok {
-		return proto.Clone(message)
-	}
-	return value
+	ctx           context.Context
+	ms            MultiStore
+	header        abci.Header
+	chainID       string
+	txBytes       []byte
+	logger        log.Logger
+	voteInfo      []abci.VoteInfo
+	gasMeter      GasMeter
+	blockGasMeter GasMeter
+	checkTx       bool
+	minGasPrice   DecCoins
+	consParams    *abci.ConsensusParams
+	evtManager    *EventManager
+}
+
+// Proposed rename, not done to avoid API breakage
+type Request = Context
+
+// Read-only accessors
+func (c Context) Context() context.Context    { return c.ctx }
+func (c Context) MultiStore() MultiStore      { return c.ms }
+func (c Context) BlockHeight() int64          { return c.header.Height }
+func (c Context) BlockTime() time.Time        { return c.header.Time }
+func (c Context) ChainID() string             { return c.chainID }
+func (c Context) TxBytes() []byte             { return c.txBytes }
+func (c Context) Logger() log.Logger          { return c.logger }
+func (c Context) VoteInfos() []abci.VoteInfo  { return c.voteInfo }
+func (c Context) GasMeter() GasMeter          { return c.gasMeter }
+func (c Context) BlockGasMeter() GasMeter     { return c.blockGasMeter }
+func (c Context) IsCheckTx() bool             { return c.checkTx }
+func (c Context) MinGasPrices() DecCoins      { return c.minGasPrice }
+func (c Context) EventManager() *EventManager { return c.evtManager }
+
+// clone the header before returning
+func (c Context) BlockHeader() abci.Header {
+	// TODO: figure out clone better
```

Comment:
Guessing this todo will get worked out if you decide to go ahead with this approach.

Reply:
yes, it paniced about trying to clone the type, not sure why...

I will ensure there are properly cloned if approved